### PR TITLE
add parallel limt for volume_stat_caculator to resolve performance bottleneck

### LINKED ISSUE #103728: kubelet volume_stat_caculator of fsResourceAnalyzer performance bottleneck while restart
#### What happened:

For single node with 1000 pods , 3 containers per pod and 3 volumes per container,  while kubelet restart, there are many volume du commands,  and the system loadaverage becomes very high, system down and not responding

```
top - 09:25:53 up 16:35,  2 users,  load average: **178.48**, 63.64, 50.28
Tasks: 8572 total, 227 running, 8329 sleeping,  15 stopped,   1 zombie
%Cpu(s):  9.5 us, 63.3 sy, 22.0 ni,  0.6 id,  0.0 wa,  0.9 hi,  3.7 si,  0.0 st,  0.6 id_exact,  0.9 hi_exa  5.0 irq_exact
KiB Mem : 49278232 total, 15596056 free, 31670396 used,  2011780 buff/cache
KiB Swap:        0 total,        0 free,        0 used. 16888856 avail Mem

   PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
**199570** root      20   0   14.4g   1.1g  14632 S 227.3  2.2   3:21.84 kubelet
  5993 root      20   0   46.5g   1.8g    824 S  38.8  3.9 327:52.19 dockerd
205595 root      20   0  144496   1864   1352 D  24.7  0.0   0:00.85 ps
205345 root      20   0   11952   1932    564 D  15.8  0.0   0:00.54 pidof
199634 root      20   0  160700  11668   1492 R  13.2  0.0   0:10.61 top
205691 root      20   0  144496   1864   1348 R  12.6  0.0   0:00.44 ps
205352 root      39  19  229132  41736    708 R   8.3  0.1   0:00.28 du
205445 root      39  19  229132  13096    708 R   8.3  0.0   0:00.28 du
205320 root      39  19  110372   3240    708 R   8.0  0.0   0:00.27 du
205331 root      39  19  229132  41892    708 R   7.8  0.1   0:00.27 du
205450 root      39  19  229132  32516    708 R   7.8  0.1   0:00.27 du
205452 root      39  19  229132  40496    708 R   8.0  0.1   0:00.27 du
205455 root      39  19  229132  41004    708 R   8.0  0.1   0:00.27 du
205489 root      39  19  229132  41716    708 R   8.0  0.1   0:00.27 du
  3041 root      20   0 2451620 142224    596 S   7.5  0.3  32:46.41 userkpi
205327 root      39  19  229132   3244    708 R   7.8  0.0   0:00.26 du
205422 root      39  19  229132  79276    704 R   7.8  0.2   0:00.26 du
205473 root      39  19  229132  96664    708 R   7.5  0.2   0:00.26 du
205478 root      39  19  229132 119480    708 R   7.5  0.2   0:00.26 du
205514 root      39  19  230152 120424    704 R   7.8  0.2   0:00.26 du
205379 root      39  19  230152 120420    704 R   7.5  0.2   0:00.25 du
205520 root      39  19  229132  93876    708 R   7.5  0.2   0:00.25 du
205528 root      39  19  229132  90068    708 R   7.5  0.2   0:00.25 du
205557 root      39  19  230152 120424    704 R   7.5  0.2   0:00.25 du
     9 root      20   0       0      0      0 S   6.6  0.0   2:43.53 rcu_sched
205277 root      39  19  110372   3244    708 R   6.6  0.0   0:00.27 du
205362 root      39  19  230152 120420    704 R   7.2  0.2   0:00.24 du
205392 root      39  19  230152 120424    704 R   6.9  0.2   0:00.24 du
```

```
root     207835 **199570**  0 09:26 ?        00:00:00 du -x -s -B 1 /docker/pods/95bc712e-e2ad-41ac-8de4-8f0fde2e5747/volumes/kubernetes.io~empty-dir/cache-volume
root     207838 199570  0 09:26 ?        00:00:00 du -x -s -B 1 /docker/pods/d074dbb5-2e02-46f7-8052-ade2a01108db/volumes/kubernetes.io~secret/default-token-6khlg
root     207843 199570  0 09:26 ?        00:00:00 du -x -s -B 1 /docker/pods/9eb62c81-7ceb-4bce-b5b5-5a717e8a2cca/volumes/kubernetes.io~empty-dir/cache-volume
root     207844 199570  0 09:26 ?        00:00:00 du -x -s -B 1 /docker/pods/9aac3861-f469-4273-a397-c0a40ea5cb0f/volumes/kubernetes.io~secret/default-token-6khlg
```
https://github.com/kubernetes/kubernetes/blob/40bc3a81a8795c45b019fd720b5f22657e9ab136/pkg/kubelet/server/stats/volume_stat_calculator.go#L135

#### What you expected to happen:

kubelet restart with low loadaverage, may limit the num of du commands parallelized

#### Environment:
- Kubernetes version (use `kubectl version`): v1.19.4
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): centos 7  
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others: VM-flavor   24C 48G

### PR DESCRIPTION
**Pull request by 249043822**
#### What type of PR is this?
/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #103728

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by endocrimes** [comment]
This isn't really a rate limiter (executions / time period), but a semaphore (bounding parallel calls). Rather than implementing one informally with blocking channels, you might want to consider using https://pkg.go.dev/golang.org/x/sync/semaphore instead.

**Comment by 249043822** [comment]
> This isn't really a rate limiter (executions / time period), but a semaphore (bounding parallel calls). Rather than implementing one informally with blocking channels, you might want to consider using https://pkg.go.dev/golang.org/x/sync/semaphore instead.
> 
> Thundering herd disk stats collection is kinda expensive, so introducing some bounding is probably a reasonable thing to do here (because there isn't really much we can do to make it significantly cheaper).

@endocrimes thanks, I have added semaphore

**Comment by cheftako** [comment]
Sorry couple of silly questions.

You've set the semaphores concurrent call limit to 100. Where did we get that value from? It wasn't clear to me from the original bug how many concurrent du instances you were dealing with. Looking at the issue 100 instances of du look like it will already be using 5-10 cores. Would it make sense for this value to be configurable?

The sync.Once is part of the fsResourceAnalyzer struct so new instance of that struct will have a seperate syncOnce and a separate semaphore. So if you have multiple fsResourceAnalyzer instances your actually concurrent call limit will be a multiple of 100.  

Also if we have multiple fsResourceAnalyzer we won't take advantage of cached values computed while we were blocked. Would it make sense to see if a some other instances have gotten a value while we were blocked?

**Comment by 249043822** [comment]
> Sorry couple of silly questions.
> 
> You've set the semaphores concurrent call limit to 100. Where did we get that value from? It wasn't clear to me from the original bug how many concurrent du instances you were dealing with. Looking at the issue 100 instances of du look like it will already be using 5-10 cores. Would it make sense for this value to be configurable?

Yea, making this value configurable may be greater. 100 is from what I have tested that restarting kubelet could be stable and the system loadaverage is within control.  Also the value  maybe arbitrary.  

> the original bug how many concurrent du instances you were dealing with
Actually I could not clearly caculate the num, because while restarting kubelet some time. the os system becomes down.  but the volume num is 1000 * 3 * 3

> The sync.Once is part of the fsResourceAnalyzer struct so new instance of that struct will have a seperate syncOnce and a separate semaphore. So if you have multiple fsResourceAnalyzer instances your actually concurrent call limit will be a multiple of 100.
> 
> Also if we have multiple fsResourceAnalyzer we won't take advantage of cached values computed while we were blocked. Would it make sense to see if a some other instances have gotten a value while we were blocked?

thanks, I didn't know the scene of multiple fsResourceAnalyzer. are multiple fsResourceAnalyzer stat the shared pod volumes all in the kubelet?

File: pkg/kubelet/server/stats/fs_resource_analyzer.go [hunk 1]
```diff
@@ -28,6 +29,9 @@ import (
 	"k8s.io/klog/v2"
 )
 
+// limit burst size
```
Comment (dashpole) [inline pkg/kubelet/server/stats/fs_resource_analyzer.go:32]: nit: comment should be of the form `// LimitSize ...`

File: pkg/kubelet/server/stats/volume_stat_calculator.go [hunk 2]
```diff
@@ -128,6 +132,13 @@ func (s *volumeStatCalculator) calcAndStoreStats() {
 		volumesSpec[v.Name] = v
 	}
 
+	// bounding parallel
+	// refer issue #103728 for detail
```
Comment (dashpole) [inline pkg/kubelet/server/stats/volume_stat_calculator.go:136]: Doesn't this bound more than just `du` calls, and apply to all volume types?

**Review by dashpole — CHANGES_REQUESTED** [review]
We should be careful here, as this could cause disk stats to stop working correctly, even for users that don't have IO problems.  Also, the bug referenced in this PR describes an issue with `du` calls, which the kubelet doesn't make anymore.  Before we take this approach, we should verify that the problem still persists  after the change: #96115.

**Comment by 249043822** [comment]
> We should be careful here, as this could cause disk stats to stop working correctly, even for users that don't have IO problems. Also, the bug referenced in this PR describes an issue with `du` calls, which the kubelet doesn't make anymore. Before we take this approach, we should verify that the problem still persists after the change: #96115.

@dashpole I will merge #96115 into my branch and verify again,  thanks

**Comment by 249043822** [comment]
> > We should be careful here, as this could cause disk stats to stop working correctly, even for users that don't have IO problems. Also, the bug referenced in this PR describes an issue with `du` calls, which the kubelet doesn't make anymore. Before we take this approach, we should verify that the problem still persists after the change: #96115.
> 
> @dashpole I will merge #96115 into my branch and verify again, thanks

 #96115 can resolve this problem, thanks @dashpole

**Pull request closed without merging**